R/score_forecasts.R

```diff
@@ -0,0 +1,110 @@
+#' Score forecasts
+#'
+#' @param forecasts required data.frame with forecasts in the format returned
+#' by load_forecasts
+#' @param truths required data.frame with forecasts in the format returned
+#' by load_truth
+#' @param desired_score_types character vector of scores to calculate; defaults to returning
+#' all available scores
+#' @param return_format string: "long" returns long format with a column for
+#' "score_name" and a column for "score_value"; "wide" returns wide format with
+#' a separate column for each score. Defaults to "wide".
+#'
+#' @return data.frame with columns model, forecast_date, location, horizon,
+#' temporal_resolution, target_variable, horizon, target_end_date, and scores:
+#' If return_format is "long", also contains columns score_name and score_value
+#' where score_name is the type of score calculated and score_value has the numeric
+#' value of the score.
+#' If return_format is "wide", each calculated score is in its own column.
+#'
+#' @export
+score_forecasts <- function(
+  forecasts,
+  truth,
+  desired_score_types,
+  return_format = "wide"
+) {
+  
+  # forecasts data.frame format
+  # columns: model, forecast_date, location, horizon, temporal_resolution,
+  #          target_variable, target_end_date, type, quantile, value
+  forecasts_colnames <- c(
+    "model", "forecast_date", "location", "horizon", "temporal_resolution",
+    "target_variable", "target_end_date", "type", "quantile", "value"
+  )
+  
+  # validate forecasts
+  # as long as forecasts contains the columns above, it will pass the check
+  if (missing(forecasts) || is.null(forecasts)) {
+    stop("Forecast dataframe missing", call. = TRUE)
+  } else if (!any(is.element(colnames(forecasts), forecasts_colnames))) {
+    stop("Forecast dataframe columns malformed", call. = TRUE)
```

Comment:
I think it would be helpful to have this error message be more specific, to have one error thrown if there was a "Required column missing" and another if "Unknown column included." (Although is the latter really a problem?)